src/librbd/ImageWatcher.cc

```diff
@@ -121,8 +131,19 @@ int ImageWatcher::unregister_watch() {
 
   cancel_async_requests(-ESHUTDOWN);
 
-  RWLock::WLocker l(m_watch_lock);
-  return m_image_ctx.md_ctx.unwatch2(m_handle);
+  int r = 0;
+  {
+    RWLock::WLocker l(m_watch_lock);
+    assert(m_watch_state != WATCH_STATE_UNREGISTERED);
+    if (m_watch_state == WATCH_STATE_REGISTERED) {
```

Comment:
Your patch fixes the crash.   I moved the ```unwatch2``` call from ```reregister_watch``` to ```handle_error``` so that it works w/o your patch (and without introducing a new state).